ddtrace/internal/writer.py

```diff
@@ -60,40 +58,142 @@ def __init__(
         shutdown_timeout=DEFAULT_TIMEOUT,
         sampler=None,
         priority_sampler=None,
+        buffer_size=20 * 1000000,
```

Comment:
why have a different buffer size?

I'd assume we would buffer up to a size, and then just manually trigger an off-cycle flush?

Reply:
seems like we could end up in a place where we are 4 seconds behind on flushing traces.

also, is this a big/overhead to add to the app? if we can get things out of memory quicker our impact on memory should be less?

Reply:
The thinking here is that traces aren't continuous, they're discrete. So we could have a situation like the following

```
[       payload size       ]
[     t1     ][t2][       t3       ]
```

If we made the buffer size equal to the payload size then we're going to drop t3. The same kind of situation could happen if the buffer was twice the payload size but we'd be sending at least the payload size amount of traces eventually.

By having a larger buffer we could also attempt to optimize payload sizes that we're sending. For example if we had

```
[          buffer          ]
[payload size][payload size]
[ t1 ][t2][  t3   ][t4]
```

We could move t4 into the payload send a payload like
```
[          buffer          ]
[payload size][payload size]
[ t1 ][t2][t4][  t3   ]
```

Obviously we'd want to prevent starvation so not sure how practical this optimization becomes... also it's NP-complete since that would be the knapsack problem 😬. 

There's also the possibility that more traces come in while we're popping from a full queue (in which case we'd also drop traces).

None of these really matter if the buffer is at full or empty so in reality it probably isn't worth complicating the implementation by keeping these two different sizes. Especially since we can introduce off-cycle sending.

Reply:
yeah, having the buffer being able to trigger a flush is a good idea.

optimizing for the most full payloads probably isn't worth the overhead. I haven't seen too many scenarios where we try to send more than 1 payload per second anyways, so would probably be optimizing for an off case.

although, if we get rid of the 1k limit it might pop up more, we might be suppressing those situations.


if we were going to optimize payloads, it would be to ensure that a trace always fits into a payload, e.g.

```
[payload size]
[ t1                  ]
```

```
[payload size][payload size]
[ t1.a       ][ t1.b  ]
```

dropping traces larger than the payload seems to be a more common issue today.